### ISSUE TITLE
bug: state table distribution of 2-phase agg leads to wrong results

### ISSUE THREAD

**Issue by BugenZhao** [risingwavelabs/risingwave#5012]
First, set the `HASH_AGG_CACHE_SIZE` to 0.

https://github.com/risingwavelabs/risingwave/blob/e8ea42c17ff383beb973d731899c2dbada4bdd1f/src/stream/src/executor/hash_agg.rs#L42-L43

Create a materialized view with 2-phase agg.

```
dev=> create table tt (v int);
CREATE_TABLE

dev=> explain create materialized view mvv as select max(v) max from tt;
                                                              QUERY PLAN                                                               
---------------------------------------------------------------------------------------------------------------------------------------
 StreamMaterialize { columns: [max], pk_columns: [] }
   StreamProject { exprs: [max(max(tt.v))] }
     StreamGlobalSimpleAgg { aggs: [sum(count), max(max(tt.v))] }
       StreamExchange { dist: Single }
         StreamHashAgg { group_key: [Vnode(tt._row_id)], aggs: [count, max(tt.v)] }
           StreamProject { exprs: [tt.v, tt._row_id, Vnode(tt._row_id)] }
             StreamTableScan { table: tt, columns: [tt.v, tt._row_id], pk: [tt._row_id], distribution: UpstreamHashShard(tt._row_id) }
(7 rows)
```

Do some inserts.
```
dev=> select * from tt;
 v 
---
 2
 1
(2 rows)

dev=> select * from mvv;
 max 
-----
   2
(1 row)
```

Flush, then delete `v = 2`.
```
dev=> delete from tt where v = 2;
DELETE 1
dev=> select * from tt;
 v 
---
 1
(1 row)

dev=> select * from mvv;
 max 
-----
   2  <- 😄
(1 row)
```

This is caused by:
- The distribution key of the local agg state store is `_row_id`, should it be singleton instead?
- The state table ignores the invalid distribution key computation, and uses the vnode `0` (why?) for reads.
- The recovery of states / deletion is not tested well.

Related:
- https://github.com/risingwavelabs/risingwave/pull/5011#issuecomment-1231926678

cc @st1page @richardchien @wcy-fdu

**Comment by fuyufjh** [comment]
We need CH-Benchmark

**Comment by BugenZhao** [comment]
> We need CH-Benchmark 🥵

Related: #5020 is also only able to be reproduced with the no-cache configuration.

**Comment by st1page** [comment]
fixing...

**Comment by st1page** [comment]
fixed #5079
